### PR TITLE
[SPARK-22916][SQL] shouldn't bias towards build right if user does not specify

### PR DESCRIPTION
**Pull request by liufengdb**
## What changes were proposed in this pull request?

When there are no broadcast hints, the current spark strategies will prefer to building the right side, without considering the sizes of the two tables. This patch added the logic to consider the sizes of the two tables for the build side. To make the logic clear, the build side is determined by two steps:

1. If there are broadcast hints, the build side is determined by `broadcastSideByHints`;
2. If there are no broadcast hints, the build side is determined by `broadcastSideBySizes`;
3. If the broadcast is disabled by the config, it falls back to the next cases.

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)
(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)

Please review http://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

**Comment by gatorsmile** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala [hunk 1]
```diff
@@ -158,45 +158,65 @@ abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
...
-      } else if (canBuildRight && canBuildLeft) {
+      } else {
         // for the last default broadcast nested loop join
         smallerSide
-      } else {
-        throw new AnalysisException("Can not decide which side to broadcast for this join")
       }
     }
 
+    private def needsBroadcastByHints(joinType: JoinType, left: LogicalPlan, right: LogicalPlan)
```
Comment (gczsjdy) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala]: Maybe using `canBroadcastByHints` more consistent with the naming convention?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala [hunk 2]
```diff
@@ -158,45 +158,65 @@ abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
...
+    }
+
+    private def needsBroadcastByConfig(joinType: JoinType, left: LogicalPlan, right: LogicalPlan)
+      : Boolean = {
+      val buildLeft = canBuildLeft(joinType) && canBroadcast(left)
+      val buildRight = canBuildRight(joinType) && canBroadcast(right)
+      buildLeft || buildRight
+    }
+
+    private def broadcastSideByConfig(joinType: JoinType, left: LogicalPlan, right: LogicalPlan)
```
Comment (gczsjdy) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala]: Is is better to use `xxxbySize`? `byConfig` might confuse people.

**Review by gatorsmile — COMMENTED** [review]
LGTM

**Comment by gatorsmile** [comment]
Thanks! Merged to master.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala [hunk 3]
```diff
@@ -158,45 +158,65 @@ abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
...
         // for the last default broadcast nested loop join
         smallerSide
-      } else {
-        throw new AnalysisException("Can not decide which side to broadcast for this join")
       }
     }
 
+    private def canBroadcastByHints(joinType: JoinType, left: LogicalPlan, right: LogicalPlan)
+      : Boolean = {
+      val buildLeft = canBuildLeft(joinType) && left.stats.hints.broadcast
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala]: if both sides have the broadcast hint, we prefer to build left, is it expected?

**Comment by cloud-fan** [comment]
a late LGTM!

**Pull request closed without merging**